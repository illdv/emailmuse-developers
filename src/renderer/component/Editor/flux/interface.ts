/**
 * Params for editor.
 * idBackEnd get when we send Entity on back end.
 * idFrontEnd use
 */
export interface IEditEntity {
  id: string;
  folderId: number;
  idFrontEnd: string;
  html: string;
  type: EntityType;
  params: IEditEntityParameter;
}

export interface IEditEntityParameter {
  [key: string]: string;
}

export enum ParamType {
  Text = 'Text',
  Html = 'Html',
}

export enum EntityType {
  Email   = 'Email',
  Snippet = 'Snippet',
  Layout  = 'Layout',
  TemporaryLayout  = 'TemporaryLayout',
}
