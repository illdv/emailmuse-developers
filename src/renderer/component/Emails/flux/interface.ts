import { IEmail } from 'src/renderer/component/Emails/flux/interfaceAPI';
import { IPagination } from 'src/renderer/common/List/interface';
import { ActionStatus, CreateAction } from 'src/renderer/flux/interface';
import { ActionCreatorsMapObject } from 'redux';
import { EmptyActionCreator } from 'redux-act';

export interface IEmailsState {
  status: ActionStatus;
  emails: IEmail[];
  selectedEmail: IEmail;
}

export interface ILoadingTemplatePayload {
  pagination: IPagination;
  templates: IEmail[];
}

// TODO: delete ?
export interface IEmailActions extends ActionCreatorsMapObject {
  // loading?: CreateAction<{ page: number, hidePreloader?: boolean, search?: string }>;
  loading: CreateAction<{s?: string}>;
  remove?: CreateAction<string>;
  save?: CreateAction<{ email: IEmail, saveAndClose: boolean }>;
  create?: CreateAction<IEmail>;
  select?: CreateAction<IEmail>;
  failure?: EmptyActionCreator;
  successfully?: CreateAction<IEmail[]>;
  createSuccess?: CreateAction<IEmail[]>;
  copy?: CreateAction<{ id: string }>;
  selectNewTemplate?: CreateAction<{ parentId: number }>;
  getEmailFromFolder: CreateAction<{ parentId: number }>;
}
