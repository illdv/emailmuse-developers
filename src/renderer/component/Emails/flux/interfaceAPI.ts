export interface IEmail {
  id: number;
  title: string;
  type: nodeType.email;
  body: string;
  folder_id: number;
  description?: string;
  preheader?: string;
  user_id?: number;
  created_at?: string;
  updated_at?: string;
  deleted_at?: string;
}

export enum nodeType {
  email = 'email',
  folder = 'folder',
}

export interface IFolderEmail {
  id: number;
  type: nodeType;
  title: string;
  updated_at: string;
  preheader?: string;
  body?: string;
  description?: string;
  folderId?: number;
}

export interface INodesResponse {
  current_page: number;
  data: any;
  path: string;
  from: number;
  to: number;
  total: number;
  last_page: number;
  first_page_url: string;
  last_page_url: string;
  per_page: number;
  prev_page_url?: any;
  next_page_url?: any;
}
