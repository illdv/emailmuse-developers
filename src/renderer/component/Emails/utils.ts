import { IEmail, IFolderEmail, nodeType } from 'src/renderer/component/Emails/flux/interfaceAPI';
import { EntityType, IEditEntity } from 'src/renderer/component/Editor/flux/interface';
import { IFolder } from 'src/renderer/component/Folder/flux/interface';

export interface INodeItem {
  id: string;
  title: string;
  description: string;
  rightText: string;
}

export function nodeToItem(node: IEmail): INodeItem {
  return {
    id: String(node.id),
    title: node.title,
    description: node.description,
    rightText: node.updated_at,
  };
}

export function emailToFolderEmail(email: IEmail): IFolderEmail {
  return {
    id: email.id,
    title: email.title || '',
    type: email.type,
    updated_at: email.updated_at,
    description: email.description || '',
  };
}

export function folderToFolderEmail(folder: IFolder): IFolderEmail {
  return {
    id: folder.id,
    title: folder.name || '',
    type: folder.type,
    updated_at: folder.updated_at,
    description: '---',
  };
}

export function folderEmailToFolder(folderEmail: IFolderEmail): IFolder {
  return {
    id: folderEmail.id,
    name: folderEmail.title,
    type: nodeType.folder,
  };
}

export function folderEmailToEntity(folderEmail: IFolderEmail): IEditEntity {
  return {
    id: folderEmail.id,
    folderId: folderEmail.folderId,
    html: folderEmail.body,
    idFrontEnd: new Date().getTime().toString(),
    type: EntityType.Email,
    params: {
      subject: folderEmail.title,
      description: folderEmail.description,
    },
  };
}

export function emailToEditEntity({ id, body, title, description, folder_id }: IEmail): IEditEntity {
  return {
    id,
    folderId: folder_id,
    html: body,
    idFrontEnd: new Date().getTime().toString(),
    type: EntityType.Email,
    params: {
      subject: title,
      description,
    },
  };
}
