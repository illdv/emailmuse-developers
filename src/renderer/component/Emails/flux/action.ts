import { createActionCreator } from 'src/renderer/flux/utils';
import { IAsyncAction2 } from 'src/renderer/flux/interface';
import { IEmail } from 'src/renderer/component/Emails/flux/interfaceAPI';

const createAsyncAction = createActionCreator('EMAILS');
const loading = createAsyncAction('LOADING');
const remove = createAsyncAction('REMOVE');
const save = createAsyncAction('SAVE');
const create = createAsyncAction('CREATE');
const select = createAsyncAction('SELECT');
const failure = createAsyncAction('FAILURE');
const successfully = createAsyncAction('SELECT_SUBJECT');
const createSuccess = createAsyncAction('CREATE_SUCCESS');
const copy = createAsyncAction('COPY');
const selectNewTemplate = createAsyncAction('SELECT_NEW_TEMPLATE');
const getEmailFromFolder = createAsyncAction('GET_FROM_FOLDER');

export interface IEmailsActions {
  loading: IAsyncAction2<{ s?: string }>;
  remove?: IAsyncAction2<{ id: number }>;
  save?: IAsyncAction2<{ email: IEmail, saveAndClose: boolean }>;
  create?: IAsyncAction2<{ email: IEmail }>;
  select?: IAsyncAction2<{ email: IEmail }>;
  failure?: IAsyncAction2;
  successfully?: IAsyncAction2<{ emails: IEmail[] }, {}>;
  createSuccess?: IAsyncAction2<{ emails: IEmail[] }, {}>;
  copy?: IAsyncAction2<{ id: string }>;
  selectNewTemplate?: IAsyncAction2<{ parentId: number }>;
  getEmailFromFolder: IAsyncAction2<{ parentId: number }, {emails: IEmail[]}>;
}

export const emailActions: IEmailsActions = {
  loading,
  remove,
  save,
  create,
  select,
  failure,
  successfully,
  createSuccess,
  copy,
  selectNewTemplate,
  getEmailFromFolder,
};
