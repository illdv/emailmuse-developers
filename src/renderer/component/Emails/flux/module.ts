import { createReducer } from 'redux-act';

import { DrawerMenuAction } from 'src/renderer/component/Menu/flux/action';
import { ActionStatus } from 'src/renderer/flux/interface';
import { emailActions } from 'src/renderer/component/Emails/flux/action';
import { IEmail } from 'src/renderer/component/Emails/flux/interfaceAPI';

export interface IEmailsState {
  status: ActionStatus;
  emails: IEmail[];
  selectedEmail: IEmail;
}

const initialState: IEmailsState = {
  emails: [],
  selectedEmail: null,
  status: ActionStatus.REQUEST,
};

const reducer = createReducer({}, initialState);

reducer.on(emailActions.save.REQUEST, state => ({
  ...state,
  lastCreatedId: null,
}));

reducer.on(emailActions.successfully.REQUEST, (state, payload): IEmailsState => ({
  ...state,
  emails: payload.emails,
  status: ActionStatus.SUCCESS,
}));

reducer.on(emailActions.failure.REQUEST, (state): IEmailsState => ({
  ...state,
  status: ActionStatus.FAILURE,
}));

reducer.on(emailActions.remove.REQUEST, (state, payload) => ({
  ...state,
  emails: state.emails.filter(email => email.id !== payload.id),
  selectedTemplate: null,
  status: ActionStatus.REQUEST,
}));

reducer.on(emailActions.createSuccess.REQUEST, (state, payload) => ({
  ...state,
  status: ActionStatus.SUCCESS,
}));

reducer.on(DrawerMenuAction.selectMenuItem, state => ({
  ...state,
  selectedTemplate: null,
}));

export default reducer;
