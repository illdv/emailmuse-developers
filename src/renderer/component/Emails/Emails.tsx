import * as React from 'react';
import { connect, Dispatch } from 'react-redux';
import { Fade, Grid, Paper, Toolbar, Typography } from '@material-ui/core';
import { Add, Close, CreateNewFolder } from '@material-ui/icons';

import { IGlobalState } from 'src/renderer/flux/rootReducers';
import { folderEmailToEntity, folderEmailToFolder, nodeToItem } from 'src/renderer/component/Emails/utils';
import { Fab } from 'src/renderer/common/Fab';
import { IEmail, IFolderEmail, nodeType } from 'src/renderer/component/Emails/flux/interfaceAPI';
import { ActionStatus } from 'src/renderer/flux/interface';
import { bindModuleAction } from 'src/renderer/flux/saga/utils';
import { EditorActions, IEditorActions } from 'src/renderer/component/Editor/flux/actions';
import { FolderActions, IFolderActions } from 'src/renderer/component/Folder/flux/actions';
import { NodeTableList } from 'src/renderer/component/Emails/NodeList/NodeTableList';
import { IFolder } from 'src/renderer/component/Folder/flux/interface';
import { Search } from 'src/renderer/common/Search';
import { IEmailsState } from 'src/renderer/component/Emails/flux/reducer';
import { emailActions, IEmailsActions } from 'src/renderer/component/Emails/flux/action';

export namespace EmailListSpace {
  export interface IProps {
    emails: IEmailsState;
    folders: IFolder[];
    emailsActions: IEmailsActions;
    foldersActions: IFolderActions;
    editorActions: IEditorActions;
    history: any;
    location: any;
  }

  export interface IState {
    newEmail: IEmail;
    currentFolder: IFolder;
    searchWorld: string;
  }
}

export class Emails extends React.Component<EmailListSpace.IProps, EmailListSpace.IState> {
  state: EmailListSpace.IState = {
    newEmail: null,
    searchWorld: '',
    currentFolder: null,
  };

  static getDerivedStateFromProps(props, prevState) {
    const { match } = props;
    const newState  = prevState;
    if (match.params.id) {
      newState.currentFolder = {
        name: match.params.name,
        id: match.params.id,
      };
    } else {
      newState.currentFolder = { name: null, id: null };
    }
    return newState;
  }

  onChangePage = (e, page: number) => {
    this.props.emailsActions.loading.REQUEST({});
  }

  handleOpenItem = (item: IFolderEmail) => {
    if (item.type === nodeType.folder) {
      const folder = folderEmailToFolder(item);
      this.props.foldersActions.openFolder.REQUEST({ folder });
    } else {
      console.log("item", item);
      const entity = folderEmailToEntity(item);
      console.log("entity", entity)
      this.props.editorActions.edit.REQUEST(entity);
    }
  }

  onSelectNewTemplate = () => {
    this.props.emailsActions.selectNewTemplate.REQUEST({ parentId: this.state.currentFolder.id });
  }

  onCreateNewFolder = () => {
    this.props.foldersActions.showModal.REQUEST({ parentId: this.state.currentFolder.id });
  }

  handleUpdateItem = (data: { id: number, folder_id: number }) => {
    this.props.emailsActions.save.REQUEST({ email: data } as any);
  }

  handleDeleteItem = (item: IFolderEmail) => {
    if (item.type === nodeType.folder) {
      this.props.foldersActions.deleteFolder.REQUEST({ ids: [item.id] });
    } else {
      this.props.emailsActions.remove.REQUEST({ id: item.id });
    }
  }

  onCopy = (id: string) => {
    this.props.emailsActions.copy.REQUEST({ id });
  }

  onSearch = (searchWorld: string) => {
    this.props.emailsActions.loading.REQUEST({ s: searchWorld });
    this.setState({ searchWorld });
  }

  setCurrentParentId = (id: number) => {
    this.setState({ currentFolder: { ...this.state.currentFolder, id } });
  }

  onOpenRootFolder = () => {
    this.setState({ currentFolder: null });
    this.props.foldersActions.openFolder.REQUEST({});
  }

  render() {
    let { emails }                       = this.props.emails;
    const { status }                     = this.props.emails;
    const folders: IFolder[]             = this.props.folders;
    const { searchWorld, currentFolder } = this.state;
    if (status === ActionStatus.FAILURE) {
      return (
        <Typography variant='headline' noWrap align='center'>
          Sorry but we couldn't download the emails.
        </Typography>
      );
    }

    if (currentFolder && currentFolder.id === null && searchWorld === '') {
      emails = emails.filter(email => email.folder_id === null);
    }

    return (
      <div>
        <Fade in timeout={1000}>
          <Paper>
            <Grid item xs={12}>
              <Grid
                container
                spacing={16}
                alignItems={'center'}
                justify={'space-between'}
                style={{ marginTop: 0 }}
              >
                <Grid item>
                  <Toolbar>
                    <Typography
                      color='inherit'
                      variant='subheading'
                      onClick={this.onOpenRootFolder}
                      style={{ cursor: 'pointer' }}
                    >Emails
                    </Typography>
                    {
                      currentFolder && currentFolder.name ?
                        <>
                          <span style={{ padding: '0 5px 0 5px' }}> / </span>
                          <Typography
                            color='inherit'
                            variant='subheading'
                          >
                            {currentFolder.name}
                          </Typography>
                        </>
                        : null
                    }
                  </Toolbar>
                </Grid>
                <Grid style={{ paddingRight: 15 }} item>
                  <Search search={this.onSearch}/>
                </Grid>
              </Grid>
            </Grid>
            <NodeTableList
              emails={emails}
              folders={folders}
              toItem={nodeToItem}
              onOpenItem={this.handleOpenItem}
              onDeleteItem={this.handleDeleteItem}
              onUpdateItem={this.handleUpdateItem}
              onChangePage={this.onChangePage}
              onCopy={this.onCopy}
              isLoading={status === ActionStatus.REQUEST}
              onCurrentParentId={this.setCurrentParentId}
            />
            {
              (currentFolder && currentFolder.id === null) ?
                <Fab
                  onClick={this.onCreateNewFolder}
                  icon={<CreateNewFolder/>}
                  position={1}
                  title={'Add a folder'}
                  whitCtrl
                  hotKey={'F'}
                />
                :
                null

            }
            {
              (currentFolder && currentFolder.id !== null) ?
                <Fab
                  onClick={this.onOpenRootFolder}
                  icon={<Close/>}
                  position={1}
                  title={'Close'}
                />
                :
                null
            }
            <Fab
              onClick={this.onSelectNewTemplate}
              icon={<Add/>}
              position={0}
              title={'Add a new email'}
              whitCtrl
              hotKey={'A'}
            />
          </Paper>
        </Fade>
      </div>
    );
  }
}

const mapStateToProps = (state: IGlobalState) => ({
  emails: state.emails,
  folders: state.folders.folders,
});

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  emailsActions: bindModuleAction(emailActions, dispatch),
  foldersActions: bindModuleAction(FolderActions, dispatch),
  editorActions: bindModuleAction(EditorActions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Emails);
