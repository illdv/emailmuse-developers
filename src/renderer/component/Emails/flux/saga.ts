import { all, call, put, takeEvery, select } from 'redux-saga/effects';
import { AxiosResponse } from 'axios';

import { FluxToast, ToastType } from 'src/renderer/common/Toast/flux/actions';
import {
  IEmail,
  nodeType,
} from 'src/renderer/component/Emails/flux/interfaceAPI';
import { errorHandler } from 'src/renderer/flux/saga/errorHandler';
import { selectFromModal } from 'src/renderer/flux/saga/utils';
import { ModalWindowType } from 'src/renderer/common/DialogProvider/flux/actions';
import { Action } from 'redux-act';
import { ILayout } from 'src/renderer/component/Layouts/flux/interface';
import { EditorActions } from 'src/renderer/component/Editor/flux/actions';
import { emailToEditEntity } from 'src/renderer/component/Emails/utils';
import { EmailAPI } from 'src/renderer/API/EmailAPI';
import { FolderAPI } from 'src/renderer/API/FolderAPI';
import {
  IFolder,
  IGetFoldersEmailsResponse,
} from 'src/renderer/component/Folder/flux/interface';
import { FolderActions } from 'src/renderer/component/Folder/flux/actions';
import { emailActions } from 'src/renderer/component/Emails/flux/action';
import { runTutorial } from 'src/renderer/component/Tutorial/flux/reducer';
import { MenuItemType } from '../../Menu/flux/interface';

function* loadingFoldersAndEmails(action: Action<{ s: string }>) {
  try {
    const response: AxiosResponse<IGetFoldersEmailsResponse> = yield call(
      FolderAPI.getFoldersAndEmails,
      action.payload.s,
    );
    const {
      emails,
      folders,
    }: { emails: IEmail[]; folders: IFolder[] } = response.data;

    yield put(FolderActions.getFolders.SUCCESS({ folders }));
    yield put(emailActions.successfully.REQUEST({ emails }));
    const { tutorial } = yield select();
    if (
      localStorage.getItem(MenuItemType.EMAILS) &&
      tutorial.name === MenuItemType.EMAILS
    ) {
      yield put(runTutorial({}));
    }
  } catch (error) {
    yield put(emailActions.failure.REQUEST({}));
  }
}

function* getEmailsFromFolders(action: Action<{ parentId: number }>) {
  const parentId = action.payload.parentId;
  try {
    const response: AxiosResponse<any> = yield call(
      FolderAPI.getEmailsInFolder,
      parentId,
    );
    const emails: IEmail[] = response.data;

    yield put(emailActions.getEmailFromFolder.SUCCESS({ emails }));
  } catch (error) {
    yield put(emailActions.getEmailFromFolder.FAILURE({}));
  }
}

function* saveTemplate(action: Action<{ email: IEmail }>) {
  try {
    yield call(EmailAPI.edit, action.payload.email);
    yield put(emailActions.loading.REQUEST({}));

    yield put(FluxToast.Actions.showToast('Email saved', ToastType.Success));
  } catch (error) {
    yield put(
      FluxToast.Actions.showToast('Failed email saved', ToastType.Error),
    );
  }
}

function* createTemplate(action: Action<{ email: IEmail }>) {
  try {
    const response = yield call(EmailAPI.create, [action.payload.email]);

    yield put(emailActions.createSuccess.REQUEST({ emails: response.data }));
    yield put(FluxToast.Actions.showToast('Email created', ToastType.Success));
  } catch (error) {
    yield call(errorHandler, error);
    yield put(
      FluxToast.Actions.showToast('Failed email created', ToastType.Error),
    );
  }
}

function* removeTemplates(action) {
  try {
    yield call(EmailAPI.remove, action.payload);
    yield put(FluxToast.Actions.showToast('Email removed', ToastType.Success));
    yield put(FolderActions.openFolder.REQUEST({}));
  } catch (error) {
    yield put(
      FluxToast.Actions.showToast('Failed email removed', ToastType.Error),
    );
  }
}

function* copyTemplates(action) {
  try {
    yield call(EmailAPI.copy, action.payload.id);

    yield put(FluxToast.Actions.showToast('Template copy', ToastType.Success));
    call(loadingFoldersAndEmails, { type: '', payload: { s: '' } });
  } catch (error) {
    yield put(
      FluxToast.Actions.showToast('Failed template copy', ToastType.Error),
    );
  }
}

function* sagaSelectNewTemplate(action) {
  const actionSelectLayout: Action<{ layout: ILayout }> = yield selectFromModal(
    ModalWindowType.SelectLayout,
  );

  const selectedLayout = actionSelectLayout.payload.layout;
  yield put(
    EditorActions.edit.REQUEST(
      emailToEditEntity({
        id: null,
        title: selectedLayout.title,
        body: selectedLayout.body,
        description: '',
        preheader: '',
        folder_id: action.payload.parentId,
        type: nodeType.email,
      }),
    ),
  );
}

function* watcher() {
  yield all([
    takeEvery(emailActions.save.REQUEST, saveTemplate),
    takeEvery(emailActions.create.REQUEST, createTemplate),
    takeEvery(emailActions.remove.REQUEST, removeTemplates),
    takeEvery(emailActions.loading.REQUEST, loadingFoldersAndEmails),
    takeEvery(emailActions.selectNewTemplate.REQUEST, sagaSelectNewTemplate),
    takeEvery(emailActions.copy.REQUEST, copyTemplates),
    takeEvery(emailActions.getEmailFromFolder.REQUEST, getEmailsFromFolders),
  ]);
}

export default [watcher];
