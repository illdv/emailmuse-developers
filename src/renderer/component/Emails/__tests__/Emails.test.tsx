import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { Emails } from 'src/renderer/component/Emails/Emails';
import { ActionStatus } from 'src/renderer/flux/interface';
import { emailToEditEntity } from 'src/renderer/component/Emails/utils';
import {
  IEmail,
  nodeType,
} from 'src/renderer/component/Emails/flux/interfaceAPI';
import { RootMock } from 'src/renderer/__mock__/rootMock';
import toJson from 'enzyme-to-json';

// jest.mock('../../../src/components/ui/ColorList');

const EmailMock = props => (
  <RootMock>
    <Emails {...props} />
  </RootMock>
);

describe('<Emails/>', () => {
  let props;
  const selectedTemplate: IEmail = {
    id: 1,
    user_id: 1,
    title: 'test title',
    type: nodeType.email,
    description: 'test description',
    body: `<table style="width: 453px; height: 45px;">
           <tbody>
           <tr>
               <td>Left Test</td>
            </tr>
         </tbody>
      </table>`,
    created_at: '',
    updated_at: '',
    deleted_at: '',
    folder_id: 0,
  };
  const emptyTemplate = {
    id: null,
    user_id: 0,
    updated_at: '',
    description: '',
    title: 'Template',
    body: '',
    created_at: '',
    deleted_at: '',
  };

  const arrayTemplates = [selectedTemplate, emptyTemplate];
  const mockSelectEmail = jest.fn();
  const mockRemoveEmail = jest.fn();
  const mockSaveEmail = jest.fn();
  const mockCreateEmail = jest.fn();
  const mockCopyTemplate = jest.fn();
  const mockLoadingTemplate = jest.fn();
  const asyncActionObject = {
    REQUEST: jest.fn(),
    SUCCESS: jest.fn(),
    FAILURE: jest.fn(),
  };

  beforeEach(() => {
    props = {
      emails: {
        status: '',
        emails: {},
        selectedTemplate: null,
      },
      folders: [],
      editorActions: {
        edit: asyncActionObject,
        save: asyncActionObject,
        close: asyncActionObject,
        remove: asyncActionObject,
        saveAndClose: asyncActionObject,
      },
      foldersActions: {
        showModal: jest.fn(),
        createFolder: jest.fn(),
        updateFolder: jest.fn(),
        deleteFolder: jest.fn(),
        getFolders: jest.fn(),
      },
      emailsActions: {
        loading: jest.fn(),
        remove: mockRemoveEmail,
        save: mockSaveEmail,
        create: mockCreateEmail,
        select: mockSelectEmail,
        failure: jest.fn(),
        copy: jest.fn(),
        createSuccess: jest.fn(),
        selectNewTemplate: jest.fn(),
        getEmailFromFolder: jest.fn(),
      },
    };
  });

  test('Waite initialize', () => {
    // spinner rotates
    props.emails.status = ActionStatus.REQUEST;
    const render = shallow(
      <RootMock>
        <Emails {...props} />
      </RootMock>,
    );

    expect(toJson(render)).toMatchSnapshot();
  });

  test('Couldn\'t download the templates', () => {
    props.emails.status = ActionStatus.FAILURE;
    const render = shallow(<EmailMock {...props} />);

    expect(render).toMatchSnapshot();
  });

  test('User click on button "Add a new email"', () => {
    props.emails.selectedEmail = undefined;
    const render = shallow(
      <RootMock>
        <Emails {...props} />
      </RootMock>,
    );

    render.find('Fab').simulate('click');

    expect(mockSelectEmail).toBeCalled();
  });

  describe('Render template list', () => {
    beforeEach(() => {
      props.templates = {
        status: 'SUCCESS',
        pagination: {
          current_page: 1,
          last_page: 7,
          per_page: 15,
          total: 2,
        },
        emails: arrayTemplates,
        selectedTemplate: null,
      };
    });
    test('show list', () => {
      const render = mount(<EmailMock {...props} />);

      expect(render).toMatchSnapshot();
    });
    test('change page', () => {
      const page = 1;
      const render = shallow(<Emails {...props} />) as any;
      render.instance().onChangePage('e', page);

      expect(mockLoadingTemplate).toBeCalled();
      expect(mockLoadingTemplate).toBeCalledWith({ page: page + 1 });
    });

    test('User click on template item row', () => {
      const render = shallow(<EmailMock {...props} />) as any;
      render.instance().selectNode(selectedTemplate)();
      const mockEmailToEditEntity = emailToEditEntity(selectedTemplate);

      expect(props.editorActions.edit.REQUEST).toBeCalled();
      expect(props.editorActions.edit.REQUEST).toBeCalledWith(
        mockEmailToEditEntity,
      );
    });
    test('User click on copy icon', () => {
      const id = 1; // template ids
      const render = shallow(<EmailMock {...props} />) as any;
      render.instance().onCopy(id);

      expect(mockCopyTemplate).toBeCalled();
      expect(mockCopyTemplate).toBeCalledWith({ id });
    });
  });
  describe('Wen user select template', () => {
    let render;
    beforeEach(() => {
      props.emails.selectedEmail = selectedTemplate;
      render = shallow(<EmailMock {...props} />);
    });
    it('user select template', () => {
      expect(props.emails.selectedTemplate).toEqual(selectedTemplate);
      expect(render).toMatchSnapshot();
    });
    /* it('user click close button', () => {
       render.instance().onClose();
       expect(mockSelectEmail).toBeCalledWith(null);
     });*/

    /*it('user click remove button', () => {
      // if it current template
      const id = props.templates.selectedTemplate.id; // 1
      render.instance().onCloseOrRemove();
      expect(mockRemoveTemplate).toBeCalledWith(id);
    });*/
    /* it('user click save button', () => {
       // if it current template
       // props.templates.selectedTemplate.ids != false
       render.instance().onSaveOrCreate(selectedTemplate, false);

       expect(mockSaveTemplate).toBeCalledWith({ template: selectedTemplate, saveAndClose: false });
     });*/
  });

  /*describe('If it new template', () => {
    // another case when  selectedTemplate.ids = false
    let render;
    beforeEach(() => {
      props.templates.selectedTemplate = selectedTemplate;
      props.templates.selectedTemplate.ids = undefined;

      render = shallow(<Emails {...props} />);
    });
    it('user click remove button', () => {
      render.instance().onCloseOrRemove();
      expect(mockSelectEmail).toBeCalledWith(null);
    });
    it('user click save button', () => {
      render.instance().onSaveOrCreate(selectedTemplate, false);

      expect(mockCreateTemplate).toBeCalledWith(selectedTemplate);
    });
  });*/
});

// Todo: test: componentDidMount
