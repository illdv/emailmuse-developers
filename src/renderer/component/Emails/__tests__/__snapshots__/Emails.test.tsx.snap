// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Emails/> Couldn't download the templates 1`] = `
<Component>
  <Emails
    editorActions={
      Object {
        "close": Object {
          "FAILURE": [MockFunction],
          "REQUEST": [MockFunction],
          "SUCCESS": [MockFunction],
        },
        "edit": Object {
          "FAILURE": [MockFunction],
          "REQUEST": [MockFunction],
          "SUCCESS": [MockFunction],
        },
        "remove": Object {
          "FAILURE": [MockFunction],
          "REQUEST": [MockFunction],
          "SUCCESS": [MockFunction],
        },
        "save": Object {
          "FAILURE": [MockFunction],
          "REQUEST": [MockFunction],
          "SUCCESS": [MockFunction],
        },
        "saveAndClose": Object {
          "FAILURE": [MockFunction],
          "REQUEST": [MockFunction],
          "SUCCESS": [MockFunction],
        },
      }
    }
    emails={
      Object {
        "emails": Object {},
        "selectedTemplate": null,
        "status": "FAILURE",
      }
    }
    emailsActions={
      Object {
        "copy": [MockFunction],
        "create": [MockFunction],
        "createSuccess": [MockFunction],
        "failure": [MockFunction],
        "getEmailFromFolder": [MockFunction],
        "loading": [MockFunction],
        "remove": [MockFunction],
        "save": [MockFunction],
        "select": [MockFunction],
        "selectNewTemplate": [MockFunction],
      }
    }
    folders={Array []}
    foldersActions={
      Object {
        "createFolder": [MockFunction],
        "deleteFolder": [MockFunction],
        "getFolders": [MockFunction],
        "showModal": [MockFunction],
        "updateFolder": [MockFunction],
      }
    }
  />
</Component>
`;

exports[`<Emails/> Waite initialize 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Emails
      editorActions={
        Object {
          "close": Object {
            "FAILURE": [MockFunction],
            "REQUEST": [MockFunction],
            "SUCCESS": [MockFunction],
          },
          "edit": Object {
            "FAILURE": [MockFunction],
            "REQUEST": [MockFunction],
            "SUCCESS": [MockFunction],
          },
          "remove": Object {
            "FAILURE": [MockFunction],
            "REQUEST": [MockFunction],
            "SUCCESS": [MockFunction],
          },
          "save": Object {
            "FAILURE": [MockFunction],
            "REQUEST": [MockFunction],
            "SUCCESS": [MockFunction],
          },
          "saveAndClose": Object {
            "FAILURE": [MockFunction],
            "REQUEST": [MockFunction],
            "SUCCESS": [MockFunction],
          },
        }
      }
      emails={
        Object {
          "emails": Object {},
          "selectedTemplate": null,
          "status": "REQUEST",
        }
      }
      emailsActions={
        Object {
          "copy": [MockFunction],
          "create": [MockFunction],
          "createSuccess": [MockFunction],
          "failure": [MockFunction],
          "getEmailFromFolder": [MockFunction],
          "loading": [MockFunction],
          "remove": [MockFunction],
          "save": [MockFunction],
          "select": [MockFunction],
          "selectNewTemplate": [MockFunction],
        }
      }
      folders={Array []}
      foldersActions={
        Object {
          "createFolder": [MockFunction],
          "deleteFolder": [MockFunction],
          "getFolders": [MockFunction],
          "showModal": [MockFunction],
          "updateFolder": [MockFunction],
        }
      }
    />
  </Router>
</Provider>
`;
